pkg/adaptor/mongodb/reader.go

```diff
@@ -0,0 +1,160 @@
+package mongodb
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/data"
+	"github.com/compose/transporter/pkg/message/ops"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+var (
+	_ client.Reader = &Reader{}
+)
+
+// Reader implements the behavior defined by client.Reader for interfacing with MongoDB.
+type Reader struct {
+	db string
+}
+
+func newReader(db string) client.Reader {
+	return &Reader{db}
+}
+
+type resultDoc struct {
+	doc bson.M
+	c   string
+}
+
+func (r *Reader) Read(filterFn client.NsFilterFunc) client.MessageChanFunc {
+	return func(s client.Session, done chan struct{}) (chan message.Msg, error) {
+		out := make(chan message.Msg)
+		session := s.(*Session)
+		go func() {
+			defer close(out)
+			defer log.With("db", r.db).Infoln("starting Read func")
+			cSession := session.mgoSession.Copy()
+			defer cSession.Close()
+			collections := r.listCollections(cSession, filterFn)
+			iSession := session.mgoSession.Copy()
+			defer iSession.Close()
+			results := r.iterateCollection(iSession, collections)
+			for {
+				select {
+				case <-done:
+					return
+				case result, ok := <-results:
+					if !ok {
+						log.With("db", r.db).Infoln("Read completed")
+						return
+					}
+					msg := From(ops.Insert, fmt.Sprintf("%s.%s", r.db, result.c), data.Data(result.doc))
+					out <- msg
+				}
+			}
+		}()
+
+		return out, nil
+	}
+}
+
+func (r *Reader) listCollections(mgoSession *mgo.Session, filterFn func(name string) bool) <-chan string {
+	out := make(chan string)
+	go func() {
```

Comment:
I'll leave this up to you, but in my opinion, it might be cleaner to have `listCollections` and `iterateCollections` take the `out` channel as a parameter and have them called like 

```go
collOutCh := make(chan string)
results := make(chan resultDoc)
go r.listCollections(cSession, filterFn, collOutCh)
go r.iterateCollection(iSession, collOutCh, results)
```

so it's a little clearer that the functions are async and the function themselves don't have to use anonymous goroutines.

Just a matter of style though.